### ISSUE TITLE
Additional configuration needed in IntelliJ IDEs?

### ISSUE THREAD

**Issue by fgblomqvist** [mrmckeb/typescript-plugin-css-modules#96]
I came here looking for a way to improve the CSS modules experience, since it seems a bit subpar in the typescript world. Back a few years ago, I was working on a css/js-only project and we would get compile-time errors (and in IDE of course) if you used a non-existant class. On top of that, you could also get errors if you had unused classes. Very nice. Now, I understand the compile-time limitation after reading some of the other issues, and that's just unfortunate.

However, what I can't wrap my head around is what this plugin adds? I already have autocompletion in my IDE (IntelliJ IDEA), just like the gif in the readme. However, what I don't have is errors for invalid/non-existent classes. Though, this plugin doesn't seem to fix/add that. Idk if something is wrong with some configs somewhere (we use CRA, rescripts version is `3.4.1`), or if my expectations are wrong.

I'm guessing the issue with not getting errors (just in the IDE) is because of the type definitions that rescripts have (and that are listed in the readme), that allow any keys (if I understand them correctly). I'm guessing there's just something I don't understand correctly, and regardless, big kudos to you for writing this plugin :)

**Comment by fgblomqvist** [comment]
Actually, I don't get the type definitions that you see when I hover over my `styles` or over a certain class, so something isn't correct with the config.

**Comment by ekilah** [comment]
Somewhat unrelated, but there are some issues I'm experiencing with this plugin in WebStorm, xposting here.  https://github.com/mrmckeb/typescript-plugin-css-modules/issues/97

**Comment by mrmckeb** [comment]
Thanks for raising this @fgblomqvist, and thanks for the feedback. The plugin will give IDE-only errors for missing classes, so you won't see them during build (a limitation of the TypeScript Language Service).

This could be an issue with IntelliJ IDEs as posted by @ekilah... The contributors are all on VSCode, so this isn't something we've tested yet sorry.

@fgblomqvist, if you can provide a small repro, I can take a look at this over the weekend and fix if possible.

**Comment by fgblomqvist** [comment]
Thanks for getting back to me. I probably won't be able to get around to this until Monday/Tuesday, but I'll investigate further then for sure!

**Comment by mrmckeb** [comment]
Thank you :) let me know how you go.

**Comment by fgblomqvist** [comment]
Started looking at this again. Turns out IntelliJ does _not_ provide autocompletion by default. I think there's some delay/cache issue when installing/uninstalling this plugin, which is why I saw autocompletion even after I uninstalled it last time.

IntelliJ provides this oob:
* If you type `styles.x` and x is declared in the CSS file, you can use "Go to definition" to go to it
  * If x is not declared, it will say that it's invalid once you try to "Go to definition"
  * It only works for `styles.x` syntax, not `styles['x']`

Now, I could have sworn that IntelliJ did a bit more than that just a few weeks ago, but idk.
I have installed your plugin again but it doesn't seem to work right away. Tried restarting the IDE without luck. Will try to invalidate caches tomorrow and see if that gets it going. Then I shall see what ends up working/not working.

**Comment by fgblomqvist** [comment]
Couldn't get the recommended TSS_LOG variable to work but remembered that there's a TypeScript tab/window in IntelliJ (usually sits at the bottom left besides Debug/TODO etc.). That one had a log of what was going on. Turns out that my IntelliJ couldn't properly find the plugin yesterday (we use lerna and it seems like you have to put the plugin the same package that the typescript is coming from, aka. the root package). Once I solved that issue I now get types generated :tada:

Issues/behavior that I see:
1. The types that are generated are all classes (both nested and non-nested) in the SCSS file, except for those that are marked as `:global` (and the name is preserved for all classes as they were declared, dashes or not)
2. Specifying a class that isn't apart of the types shows an error (perfect)
3. Typing `styles.` only gives me autocompletion for non-dashed classes
4. Typing `styles['<some letter>` gives me autocompletion for all (matching) classes, both nested and non-nested, **as well as** ones marked as `:global`
    1. However, the autocompletion for dashed classes is their camelcase version (so if I have a class declared as `my-btn` it will be listed in autocompletion as `myBtn`).
5. Go to declaration only works for camelcase classes
    1. Oddly enough, if I have a class `my-btn` and type it in the tsx file as `myBtn` (which will show an error), I can go to its declaration without issues.

It seems like IntelliJ has assumed control over the autocompletion list as well as the declaration linkage. It also seems like IntelliJ assumes that you have the setting `exportLocalsConvention: 'camelCaseOnly',` for your CSS modules (which is not true for CRA, even if I would want it to be lol).

Now, I still get errors if I choose a class that doesn't exist, or if I go with the wrong name of one (e.g. if I go with `myBtn` as suggested, even though it's exported as `my-btn`), which is very useful. However, it would indeed be nice if the autocompletion and go-to-declaration were correct. IntelliJ is clearly aware of the types that are generated (if I hold ctrl and hover over `styles` it'll show the types that your plugin generated), so idk why it's not relying on those for autocompletion. I'm guessing I should file a bug with them about that? Or do you have any idea why that may be the case?

So yeah, once the above issue has been solved, it should be possible to see if there are any remaining problems/annoyances. I guess the one thing that would be nice is if `:global` classes weren't ignored when generating the types. But I'm guessing that could be fixed (or perhaps worked around using your custom stuff?).

**Comment by fgblomqvist** [comment]
Found the relevant bugs:
https://youtrack.jetbrains.com/issue/WEB-38105    <- Don't assume `camelCaseOnly` setting
https://youtrack.jetbrains.com/issue/WEB-41304    <- Make use of the types generated by this plugin (sorta)

**Comment by mrmckeb** [comment]
Thanks for the updates @fgblomqvist, and sorry for the slow reply - I've been away and didn't get back to this as soon as I had planned.

I'm glad that you have it mostly resolved now. I'd love some help resolving any of the remaining issues, if you feel you have the time to contribute.

I think global classes might be harder to solve, as this plugin traverses imported files, and it wouldn't be able to catch global classes declared in fildes that aren't imported. This could be solved by changing the renderer (you can [create custom renderers](https://github.com/mrmckeb/typescript-plugin-css-modules#customrenderer) too) to look at all files of that type in the base folder... but I feel that could cause a performance impact.

It's a little harder to support products from JetBrains, as they're not open source and I also don't own them. This plugin relies directly on the TypeScript Language Service, which is part of TypeScript, and baked into VSCode.

**Comment by ekilah** [comment]
@fgblomqvist can you show screenshots of the plugin working in your JetBrains editor, so I can see more about what it adds for you? I haven't seen this plugin add anything to the default behavior already present in WebStorm yet. I know what it does for my coworkers in VSCode, though.

Regarding your workaround / fix to get things going:

> Turns out that my IntelliJ couldn't properly find the plugin yesterday (we use lerna and it seems like you have to put the plugin the same package that the typescript is coming from, aka. the root package).

I'm not sure what package you're referring to or what you changed. This may or may not be relevant to my use case?

**Comment by fgblomqvist** [comment]
**tl;dr it provides an error message/indicator when you try to use a class that doesn't exist**

Holding Ctrl and hovering over the imported styles gives me a proper type spec:
<img width="160" alt="Screen Shot 2020-09-21 at 9 38 01 AM" src="https://user-images.githubusercontent.com/649260/93773465-2e8cf780-fbee-11ea-890e-d0a9d2b6f82b.png">

Trying to use a class that does not exist gives me an error in the editor:
<img width="737" alt="Screen Shot 2020-09-21 at 9 37 44 AM" src="https://user-images.githubusercontent.com/649260/93773459-2b920700-fbee-11ea-80ab-a49066b6cde5.png">

That's what this plugin adds. IntelliJ only provides autocompletion support, however, it's faulty (see links I posted further up) so I wish it could be turned off because it takes precedence over the types generated by this plugin. Once they fix that autocompletion it should go hand-in-hand with this plugin and the experience will be pretty great.

**Comment by ekilah** [comment]
yeah, now that #99 is fixed, I get those too, using TS 4.0.2.
![image](https://user-images.githubusercontent.com/3213339/93799606-39e32180-fbf4-11ea-804f-768e73763bbc.png)
![image](https://user-images.githubusercontent.com/3213339/93799656-49fb0100-fbf4-11ea-9e2a-1b6f309b7d6a.png)
![image](https://user-images.githubusercontent.com/3213339/93799700-5da66780-fbf4-11ea-95c6-98e85f1f18f2.png)

**Comment by fgblomqvist** [comment]
@mrmckeb Thanks for your response, no worries on the time.

The global classes issue I mentioned is not an actual issue (after thinking about it more). A global class shouldn't be referenced like this: `className={styles.myclass}` but rather `className="myclass"` because after all, they're global.

So yeah, the only remaining issues are strictly related to IntelliJ's implementation and the bugs I linked above. Nothing to do about it until they are solved. Will close this for now I suppose.

**Comment by mrmckeb** [comment]
Thanks for the updates everyone!
